core/trino-main/src/test/java/io/trino/sql/planner/TestDomainTranslator.java

```diff
@@ -1083,6 +1083,23 @@ public void testPredicateWithVarcharCastToDate()
                 and(
                         greaterThanOrEqual(cast(C_VARCHAR, DATE), new GenericLiteral("DATE", "2001-01-31")),
                         lessThanOrEqual(cast(C_VARCHAR, DATE), new GenericLiteral("DATE", "2005-09-10"))));
+
+        assertPredicateTranslates(
+                between(new GenericLiteral("DATE", "2001-01-31"), cast(C_VARCHAR, DATE), cast(C_VARCHAR_1, DATE)),
```

Comment:
the bug is reproducible with a simpler case

```java
        assertPredicateTranslates(
                greaterThan(new GenericLiteral("DATE", "2001-01-31"), cast(C_VARCHAR, DATE)),
                tupleDomain(
                        C_VARCHAR,
                        Domain.create(ValueSet.ofRanges(
                                Range.lessThan(VARCHAR, utf8Slice("2002")),
                                Range.greaterThan(VARCHAR, utf8Slice("9"))),
                                false)),
                greaterThan(new GenericLiteral("DATE", "2001-01-31"), cast(C_VARCHAR, DATE)));
```

- please add a simpler case
- please keep the BETWEEN case, as the BETWEEN explains why this cases needs to be handled correctly (and not eg rejected)
- please put some comment on both, for consistently with the rest of the method
  - can be as short as "regression test for ..."